File: src/comm_debugger/transports/tcp.py
from __future__ import annotations

import asyncio
from contextlib import suppress
from typing import Any

from comm_debugger.events import EventBroker
from comm_debugger.models import TcpClientConfig, TcpServerConfig
from comm_debugger.transports.base import BaseTransport, TransportError


def _peer_name(writer: asyncio.StreamWriter) -> str:
    peer = writer.get_extra_info("peername")
    if isinstance(peer, tuple) and len(peer) >= 2:
        return f"{peer[0]}:{peer[1]}"
    return str(peer or "未知客户端")


class TcpClientTransport(BaseTransport):
    mode = "tcp_client"

    def __init__(self, config: TcpClientConfig, broker: EventBroker) -> None:
        super().__init__(broker)
        self.config = config
        self._reader: asyncio.StreamReader | None = None
        self._writer: asyncio.StreamWriter | None = None
        self._reader_task: asyncio.Task[None] | None = None
        self._write_lock = asyncio.Lock()

    async def start(self) -> None:
        try:
            self._reader, self._writer = await asyncio.wait_for(
                asyncio.open_connection(self.config.host, self.config.port),
                timeout=self.config.connect_timeout,
            )
        except (TimeoutError, OSError) as exc:
            self._reader = None
            self._writer = None
            raise TransportError(
                f"无法连接 TCP {self.config.host}:{self.config.port}：{exc}"
            ) from exc

        self.connected = True
        self._reader_task = asyncio.create_task(self._read_loop(), name="tcp-client-reader")
        self.publish_notice(f"已连接 TCP {self.config.host}:{self.config.port}")
        self.publish_status()

    async def stop(self) -> None:
        self.connected = False
        writer, self._writer = self._writer, None
        self._reader = None
        if writer is not None:
            writer.close()
            with suppress(Exception):
                await writer.wait_closed()

        task, self._reader_task = self._reader_task, None
        if task is not None and task is not asyncio.current_task():
            task.cancel()
            with suppress(asyncio.CancelledError):
                await task
        self.publish_status()

    async def send(self, data: bytes) -> None:
        writer = self._writer
        if not self.connected or writer is None or writer.is_closing():
            raise TransportError("TCP Client 尚未连接")
        if not data:
            return
        try:
            async with self._write_lock:
                writer.write(data)
                await writer.drain()
        except (ConnectionError, OSError) as exc:
            raise TransportError(f"TCP 发送失败：{exc}") from exc
        self.publish_data("tx", data, f"{self.config.host}:{self.config.port}")

    def details(self) -> dict[str, Any]:
        return {"host": self.config.host, "port": self.config.port}

    async def _read_loop(self) -> None:
        assert self._reader is not None
        try:
            while self.connected:
                data = await self._reader.read(65536)
                if not data:
                    break
                self.publish_data("rx", data, f"{self.config.host}:{self.config.port}")
        except asyncio.CancelledError:
            raise
        except (ConnectionError, OSError) as exc:
            if self.connected:
                self.publish_error(f"TCP 连接读取失败：{exc}")
        finally:
            if self.connected:
                self.connected = False
                self.publish_notice("TCP 远端已断开")
                self.publish_status()


class TcpServerTransport(BaseTransport):
    mode = "tcp_server"

    def __init__(self, config: TcpServerConfig, broker: EventBroker) -> None:
        super().__init__(broker)
        self.config = config
        self._server: asyncio.Server | None = None
        self._clients: dict[asyncio.StreamWriter, asyncio.Task[None]] = {}
        self._write_lock = asyncio.Lock()
        self._bound_host = config.host
        self._bound_port = config.port

    async def start(self) -> None:
        try:
            self._server = await asyncio.start_server(
                self._accept_client,
                self.config.host,
                self.config.port,
            )
        except OSError as exc:
            raise TransportError(
                f"无法监听 TCP {self.config.host}:{self.config.port}：{exc}"
            ) from exc

        sockets = self._server.sockets or []
        if sockets:
            address = sockets[0].getsockname()
            self._bound_host = str(address[0])
            self._bound_port = int(address[1])
        self.connected = True
        self.publish_notice(f"TCP Server 正在监听 {self._bound_host}:{self._bound_port}")
        self.publish_status()

    async def stop(self) -> None:
        self.connected = False
        server, self._server = self._server, None
        if server is not None:
            server.close()
            await server.wait_closed()

        clients = tuple(self._clients.items())
        self._clients.clear()
        for writer, _ in clients:
            writer.close()
        for writer, task in clients:
            with suppress(Exception):
                await writer.wait_closed()
            if task is not asyncio.current_task():
                task.cancel()
                with suppress(asyncio.CancelledError):
                    await task
        self.publish_status()

    async def send(self, data: bytes) -> None:
        if not self.connected:
            raise TransportError("TCP Server 尚未启动")
        clients = [writer for writer in self._clients if not writer.is_closing()]
        if not clients:
            raise TransportError("当前没有 TCP 客户端连接")
        if not data:
            return

        failed: list[asyncio.StreamWriter] = []
        async with self._write_lock:
            for writer in clients:
                try:
                    writer.write(data)
                    await writer.drain()
                except (ConnectionError, OSError):
                    failed.append(writer)
        for writer in failed:
            await self._remove_client(writer)
        if len(failed) == len(clients):
            raise TransportError("向所有 TCP 客户端发送均失败")
        self.publish_data("tx", data, f"广播至 {len(clients) - len(failed)} 个客户端")

    def details(self) -> dict[str, Any]:
        return {
            "host": self._bound_host,
            "port": self._bound_port,
            "client_count": len(self._clients),
            "clients": [_peer_name(writer) for writer in self._clients],
        }

    async def _accept_client(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
    ) -> None:
        task = asyncio.current_task()
        assert task is not None
        self._clients[writer] = task
        peer = _peer_name(writer)
        self.publish_notice(f"TCP 客户端已连接：{peer}")
        self.publish_status()
        try:
            while self.connected:
                data = await reader.read(65536)
                if not data:
                    break
                self.publish_data("rx", data, peer)
        except asyncio.CancelledError:
            raise
        except (ConnectionError, OSError) as exc:
            if self.connected:
                self.publish_error(f"TCP 客户端 {peer} 读取失败：{exc}")
        finally:
            await self._remove_client(writer)

    async def _remove_client(self, writer: asyncio.StreamWriter) -> None:
        if writer not in self._clients:
            return
        self._clients.pop(writer, None)
        peer = _peer_name(writer)
        writer.close()
        with suppress(Exception):
            await writer.wait_closed()
        self.publish_notice(f"TCP 客户端已断开：{peer}")
        self.publish_status()

File: server/src/transports/tcp.ts
import net, { type Server, type Socket } from "node:net";
import type { EventBroker } from "../core/event-broker.js";
import { BaseTransport, TransportError } from "../core/transport.js";
import type { TcpClientConfig, TcpServerConfig } from "../core/types.js";

function peerName(socket: Socket): string {
  return `${socket.remoteAddress ?? "未知客户端"}:${socket.remotePort ?? "?"}`;
}

const TCP_RECONNECT_INTERVAL_MS = 3_000;

export class TcpClientTransport extends BaseTransport {
  readonly mode = "tcp_client" as const;
  #socket: Socket | undefined;
  #reconnectTimer: NodeJS.Timeout | undefined;
  #reconnecting = false;
  #stopping = true;
  #generation = 0;

  constructor(
    readonly config: TcpClientConfig,
    broker: EventBroker,
  ) {
    super(broker);
  }

  async start(): Promise<void> {
    this.#stopping = false;
    this.#generation += 1;
    try {
      await this.#connect(false, this.#generation);
    } catch (error) {
      this.#stopping = true;
      throw new TransportError(
        `无法连接 TCP ${this.peer}：${errorMessage(error)}`,
        { cause: error },
      );
    }
  }

  async #connect(reconnecting: boolean, generation: number): Promise<void> {
    const socket = new net.Socket();
    this.#socket = socket;
    socket.setNoDelay(true);

    try {
      await new Promise<void>((resolve, reject) => {
        const timeout = setTimeout(() => {
          socket.destroy();
          reject(new Error("连接超时"));
        }, this.config.connect_timeout * 1_000);
        const onError = (error: Error): void => {
          clearTimeout(timeout);
          reject(error);
        };
        socket.once("error", onError);
        socket.connect(this.config.port, this.config.host, () => {
          clearTimeout(timeout);
          socket.off("error", onError);
          resolve();
        });
      });
    } catch (error) {
      if (this.#socket === socket) this.#socket = undefined;
      socket.destroy();
      throw error;
    }

    if (this.#stopping || generation !== this.#generation) {
      if (this.#socket === socket) this.#socket = undefined;
      socket.destroy();
      return;
    }

    socket.on("data", (data) => this.publishData("rx", data, this.peer));
    socket.on("error", (error) => {
      if (this.connected && this.#socket === socket) {
        this.publishError(`TCP 连接错误：${error.message}`);
      }
    });
    socket.on("close", () => this.#handleClose(socket));
    this.connected = true;
    this.#reconnecting = false;
    this.publishNotice(`${reconnecting ? "已重新连接" : "已连接"} TCP ${this.peer}`);
    this.publishStatus();
  }

  async stop(): Promise<void> {
    this.#stopping = true;
    this.#generation += 1;
    clearTimeout(this.#reconnectTimer);
    this.#reconnectTimer = undefined;
    this.#reconnecting = false;
    this.connected = false;
    const socket = this.#socket;
    this.#socket = undefined;
    if (socket && !socket.destroyed) {
      await new Promise<void>((resolve) => {
        socket.once("close", resolve);
        socket.destroy();
      });
    }
    this.publishStatus();
  }

  async send(data: Buffer): Promise<void> {
    const socket = this.#socket;
    if (!this.connected || !socket || socket.destroyed) {
      throw new TransportError("TCP Client 尚未连接");
    }
    if (data.length === 0) return;
    try {
      await new Promise<void>((resolve, reject) => {
        socket.write(data, (error) => error ? reject(error) : resolve());
      });
    } catch (error) {
      throw new TransportError(`TCP 发送失败：${errorMessage(error)}`, { cause: error });
    }
    this.publishData("tx", data, this.peer);
  }

  override details(): Record<string, unknown> {
    return {
      host: this.config.host,
      port: this.config.port,
      auto_reconnect: this.config.auto_reconnect,
      reconnecting: this.#reconnecting,
      reconnect_interval_ms: TCP_RECONNECT_INTERVAL_MS,
    };
  }

  private get peer(): string {
    return `${this.config.host}:${this.config.port}`;
  }

  #handleClose(socket: Socket): void {
    if (this.#socket !== socket) return;
    this.#socket = undefined;
    if (this.#stopping) return;
    const wasConnected = this.connected;
    this.connected = false;
    if (wasConnected) this.publishNotice("TCP 远端已断开");
    if (this.config.auto_reconnect) {
      this.#scheduleReconnect();
    } else {
      this.publishStatus();
    }
  }

  #scheduleReconnect(): void {
    if (this.#stopping || this.connected || this.#reconnectTimer) return;
    const wasReconnecting = this.#reconnecting;
    this.#reconnecting = true;
    if (!wasReconnecting) {
      this.publishNotice(`将在 ${TCP_RECONNECT_INTERVAL_MS / 1_000} 秒后重连 TCP ${this.peer}`);
      this.publishStatus();
    }
    const generation = this.#generation;
    this.#reconnectTimer = setTimeout(() => {
      this.#reconnectTimer = undefined;
      void this.#retryConnect(generation);
    }, TCP_RECONNECT_INTERVAL_MS);
  }

  async #retryConnect(generation: number): Promise<void> {
    if (this.#stopping || generation !== this.#generation) return;
    this.publishNotice(`正在重连 TCP ${this.peer}`);
    try {
      await this.#connect(true, generation);
    } catch (error) {
      if (this.#stopping || generation !== this.#generation) return;
      this.publishNotice(`TCP 重连失败：${errorMessage(error)}`);
      this.#scheduleReconnect();
    }
  }
}

export class TcpServerTransport extends BaseTransport {
  readonly mode = "tcp_server" as const;
  #server: Server | undefined;
  readonly #clients = new Set<Socket>();
  #boundHost: string;
  #boundPort: number;

  constructor(
    readonly config: TcpServerConfig,
    broker: EventBroker,
  ) {
    super(broker);
    this.#boundHost = config.host;
    this.#boundPort = config.port;
  }

  async start(): Promise<void> {
    const server = net.createServer((socket) => this.#acceptClient(socket));
    this.#server = server;
    server.on("error", (error) => {
      if (this.connected) this.publishError(`TCP Server 监听错误：${error.message}`);
    });
    try {
      await new Promise<void>((resolve, reject) => {
        server.once("error", reject);
        server.listen(this.config.port, this.config.host, () => {
          server.off("error", reject);
          resolve();
        });
      });
    } catch (error) {
      this.#server = undefined;
      throw new TransportError(
        `无法监听 TCP ${this.config.host}:${this.config.port}：${errorMessage(error)}`,
        { cause: error },
      );
    }
    const address = server.address();
    if (address && typeof address !== "string") {
      this.#boundHost = address.address;
      this.#boundPort = address.port;
    }
    this.connected = true;
    this.publishNotice(`TCP Server 正在监听 ${this.#boundHost}:${this.#boundPort}`);
    this.publishStatus();
  }

  async stop(): Promise<void> {
    this.connected = false;
    for (const socket of this.#clients) socket.destroy();
    this.#clients.clear();
    const server = this.#server;
    this.#server = undefined;
    if (server?.listening) {
      await new Promise<void>((resolve) => server.close(() => resolve()));
    }
    this.publishStatus();
  }

  async send(data: Buffer): Promise<void> {
    if (!this.connected) throw new TransportError("TCP Server 尚未启动");
    const clients = [...this.#clients].filter((socket) => !socket.destroyed);
    if (clients.length === 0) throw new TransportError("当前没有 TCP 客户端连接");
    if (data.length === 0) return;
    const results = await Promise.allSettled(clients.map((socket) => new Promise<void>((resolve, reject) => {
      socket.write(data, (error) => error ? reject(error) : resolve());
    })));
    const succeeded = results.filter((result) => result.status === "fulfilled").length;
    results.forEach((result, index) => {
      if (result.status === "rejected") clients[index]?.destroy();
    });
    if (succeeded === 0) throw new TransportError("向所有 TCP 客户端发送均失败");
    this.publishData("tx", data, `广播至 ${succeeded} 个客户端`);
  }

  override details(): Record<string, unknown> {
    return {
      host: this.#boundHost,
      port: this.#boundPort,
      client_count: this.#clients.size,
      clients: [...this.#clients].map(peerName),
    };
  }

  #acceptClient(socket: Socket): void {
    socket.setNoDelay(true);
    this.#clients.add(socket);
    const peer = peerName(socket);
    this.publishNotice(`TCP 客户端已连接：${peer}`);
    this.publishStatus();
    socket.on("data", (data) => this.publishData("rx", data, peer));
    socket.on("error", (error) => {
      if (this.connected) this.publishError(`TCP 客户端 ${peer} 通信错误：${error.message}`);
    });
    socket.on("close", () => {
      if (!this.#clients.delete(socket)) return;
      if (this.connected) {
        this.publishNotice(`TCP 客户端已断开：${peer}`);
        this.publishStatus();
      }
    });
  }
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
